src/application/stateMachine.ts

```diff
@@ -0,0 +1,135 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type NoInfer<T> = [T][T extends any ? 0 : never];
+
+interface Machine<
+  State extends string,
+  EventName extends string,
+  Context extends Record<string, unknown>,
+> {
+  send: (event: Event<EventName, Context>) => void;
+  getState: () => CurrentState<State, Context>;
+  subscribe: (listener: Listener<State, EventName, Context>) => () => void;
+  onTransition: (state: State, listener: () => void) => () => void;
+  matches: (state: State) => boolean;
+}
+
+type MachineConfig<
+  State extends string,
+  EventName extends string,
+  Context extends Record<string, unknown>,
+> = {
+  initial: NoInfer<State>;
```

Comment:
I'm curious about the decision to use `NoInfer` in the `MachineConfig` type. I say that because I find it easier to reason about when it's strongly typed since the keys and values are so similar, and it would also be a big maintainability win to have the type system catch typos where they're reversed like e.g. `devtoolsMachine.send({ type: "timedout" });` (should be `timeout`).

I see one downside re: `current.value` being inferred as `initial` though its value changes at runtime, so maybe there's a middle ground here, but curious about other challenges you maybe already ran into. A quick idea where the screaming case consts are strings:

```ts
const states = [
  INITIALIZED,
  CONNECTED,
  TIMED_OUT,
  NOT_FOUND,
  DISCONNECTED,
  RETRYING,
] as const;

const events = [CONNECT, TIMEOUT, CLIENT_NOT_FOUND, DISCONNECT, RETRY] as const;

type StatesUnionType = (typeof states)[number];
type EventsUnionType = (typeof events)[number];

interface DevToolsMachine<Context extends Record<string, unknown>> {
  initialContext?: Context;
  initial: typeof INITIALIZED;
  types: {
    events: { type?: EventsUnionType };
  };
  states: {
    [K in StatesUnionType]: {
      events?:
        | { [CONNECT]: typeof CONNECTED }
        | { [TIMEOUT]: typeof TIMED_OUT }
        | { [CLIENT_NOT_FOUND]: typeof NOT_FOUND }
        | { [DISCONNECT]: typeof DISCONNECTED }
        | { [RETRY]: typeof RETRYING };
    };
  };
}
```

Reply:
@alessbell the states can be inferred, but they are instead inferred from the rest of the object (the keys under `states`). If I allow `State` to be inferred from `initial`, it interferes with the type inference and it narrows the type of `State` to just the one declared value, rather than the union of all of them.

Here is a screenshot to show you the differences. Everything is type safe as you would expect otherwise (i.e. I can't declare an event in `events` that point to a state that doesn't exist 🙂)

**Without `NoInfer`**
<img width="756" alt="Screenshot 2024-02-08 at 9 27 04 AM" src="https://github.com/apollographql/apollo-client-devtools/assets/565661/a5b474a0-fdb1-4e19-83b4-d2c80dc43082">
<img width="760" alt="Screenshot 2024-02-08 at 9 27 10 AM" src="https://github.com/apollographql/apollo-client-devtools/assets/565661/ce721e3f-296d-41de-877f-72b89c38ab1c">

**With `NoInfer<State>`**
<img width="759" alt="Screenshot 2024-02-08 at 9 27 20 AM" src="https://github.com/apollographql/apollo-client-devtools/assets/565661/d44bbfa6-7008-4fc3-bd81-051b381bb881">
<img width="762" alt="Screenshot 2024-02-08 at 9 27 27 AM" src="https://github.com/apollographql/apollo-client-devtools/assets/565661/a947d72b-0e1f-4e8a-837c-eff9a0216d4c">

Does that help?

Reply:
And to add to this on the event types themselves, these are also type safe. Here is an example where my event is "onnected" (I deleted the "c") and you can see that TypeScript complains in my machine config:
<img width="766" alt="Screenshot 2024-02-08 at 9 33 18 AM" src="https://github.com/apollographql/apollo-client-devtools/assets/565661/3b801980-75fe-447e-b1e9-eccd3e55781a">

On the side of `machine.send(...)`, you also get the right event types listed here based on the `types` config:

<img width="588" alt="Screenshot 2024-02-08 at 9 32 35 AM" src="https://github.com/apollographql/apollo-client-devtools/assets/565661/865d097a-cb62-4dc5-b2be-84ba64fe5595">

Reply:
Ahh yep that makes sense! I see all the inference magic now :) We may want to add a comment about that since it's not immediately clear that the casting here

```
events: {} as
  | { type: "connect" }
  | { type: "timeout" }
  | { type: "disconnect" }
  | { type: "clientNotFound" }
  | { type: "retry" },
```

is responsible for typing the union of keys below.

Reply:
Ah ya I can do that. I totally stole this from [xstate](https://stately.ai/docs/machines#specifying-types) and thought it was a clever way for the types to be defined.

Reply:
Added in [6506ac8](https://github.com/apollographql/apollo-client-devtools/pull/1220/commits/6506ac89bb4a9c377249c308facaed1cff51502c)